### PR TITLE
 Issue #13155: Fix false negative in VariableDeclarationDistanceUsage 

### LINKED ISSUE #11335: False Negative: VariableDeclarationUsageDistance
I have read check documentation: https://checkstyle.sourceforge.io/config_coding.html#VariableDeclarationUsageDistance
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac YOUR_FILE.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance"/>
    </module>
</module>

/var/tmp $ cat Test.java
public class Test {
    public Test test(Object a) {
        int pc = 0; // expected violation
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test(pc).test("abc");
        return this;
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```
---

**Describe what you expect in detail.**

```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:3:9: Distance between variable 'pc' declaration and its first usage is 7
, but allowed 3.  Consider making that variable final if you still need to store 
its value in advance (before method calls that might have side effects on the original value).
 [VariableDeclarationUsageDistance]
Audit done.
Checkstyle ends with 1 errors.
```
---

Works perfectly fine with following piece of code-
```java
public class Test {
    public Test test(Object a) {
        int pc = 0; // expected violation, got a violation
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test(pc);
        return this;
    }
}
```

### LINKED ISSUE #13155: update documentation: VariableDeclarationUsageDistance on  block of initialization methods
I have read check documentation: https://checkstyle.sourceforge.io/config_coding.html#VariableDeclarationUsageDistance
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance">
            <property name="allowedDistance" value="1"/>
        </module>
    </module>
</module>

/var/tmp $ cat Test.java
public class Test {
    public Test test(Object a) {
        Test obj = test("abc");
        int var = 12; // violation expected 'Dist.* between variable 'var'.* and its first usage is 4'
        obj.test("abc");
        obj.test("abc");
        obj.test("abc");
        int k = var + 3;
        return this;
    }

    Test fooFaaHooHaa() {
        int pc = 0; // violation expected, method chain is broken
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");

        fooFaaHooHaa().test("abc");

        test(pc).test("abc");
        return null;
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.2-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

---

**Describe what you expect in detail.**
```bash
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.x-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/vyom/IdeaProjects/Config-Checkstyle/src/Test.java:4:9: Distance between variable 'var' declaration and its first usage is 4, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]
[ERROR] /home/vyom/IdeaProjects/Config-Checkstyle/src/Test.java:13:9: Distance between variable 'pc' declaration and its first usage is 8, but allowed 1.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]
Audit done.
Checkstyle ends with 2 errors.
```

There is unfinished work at https://github.com/checkstyle/checkstyle/pull/11672 and can be reused.

-------

Issue is fixed in some other update, no violations happening any more, but documentation should be updated to show example on how it works.

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
Resolves #13155

Diff Regression config: https://gist.githubusercontent.com/Vyom-Yadav/98dceb63a79f4833e85fff9b2e1464a6/raw/a875094801a518055bde1c71fa360f703d1ea7f6/my_checks.xml
Report label: AllConfig

---

Final Report: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/a99d8b4_2023120550/reports/diff/index.html

---

Coverage for FullIdent, will be incomplete for now, complete coverage is in #11231

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java
Comment (nick-mancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:489]:
```suggestion
                        final boolean isMethodCalledWithInstance =
                                methodCallIdent.getParent().getType() == TokenTypes.DOT;
```

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:489]: That won't be logically correct, I looped my way up top as we can have `TYPECAST` as the parent before we can get to `DOT`, multiple typecasts are also possible, here is a simple example-
```java
((subClass)obj).setVal1(1);
```
Related AST-
```
 |--EXPR -> EXPR
 |   `--METHOD_CALL -> (
 |       |--DOT -> .
 |       |   |--LPAREN -> (
 |       |   |--TYPECAST -> (
 |       |   |   |--TYPE -> TYPE
 |       |   |   |   `--IDENT -> subClass
 |       |   |   |--RPAREN -> )
 |       |   |   `--IDENT -> obj -------> This has parent as typecast
 |       |   |--RPAREN -> )
 |       |   `--IDENT -> setVal1
 |       |--ELIST -> ELIST
 |       |   `--EXPR -> EXPR
 |       |       `--NUM_INT -> 1
 |       `--RPAREN -> )
 |--SEMI -> ;
```

> Additionally, it might help to refactor this method a bit, as a separate commit before adding new code.

That is a good point, there is some logical simplification that is possible here, I will send a minor PR for that.

Comment (nick-mancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:489]: > That won't be logically correct, I looped my way up top…

Do we have a test case like this?

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java]: @nick-mancuso Thinking on your suggestion, all method calls without instance, their `IDENT` always has `METHOD_CALL` as its parent, so the reverse condition worked.
```java
final boolean isMethodCalledWithInstance =
                            methodCallIdent.getParent().getType() != TokenTypes.METHOD_CALL;
```
But the problem is still the same, if I negate `!=` to `==` then also all the test cases will pass, branches are unique in that case too :(. Interesting mutation, any ideas?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java
Comment (nick-mancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:504]: Try this:

```suggestion
                    if (isFirstMethodCallToBeExamined) {
                        isInitialMethodCalledWithInstance = isMethodCalledWithInstance;
                        isFirstMethodCallToBeExamined = false;
                    }

                    result = isMethodCalledWithInstance == isInitialMethodCalledWithInstance;
```

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java]: Good suggestion, implemented.

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Comment by Vyom-Yadav** [comment]
> DefaultConfig: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/87ec68d_2022080049/reports/diff/index.html

Some fixes are required, I am working on it.

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Comment by Vyom-Yadav** [comment]
@romani @nick-mancuso One pattern I would like to discuss is-
```java
int c = 12;
assertEquals(1,2);
assertEquals(1,2);
int l = 13;
assertEquals(1,l);
assertEquals(1,l);
int b = 12;
assertEquals(b,c); // c is used here
```
Now it is not a continuous chain of `assertEquals` but still no violations will be given out as we ignore variable declarations in between, is that fine?

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Review by pbludov — APPROVED** [review]

**Comment by Vyom-Yadav** [comment]
@pbludov Please assign the next reviewer.

**Comment by nick-mancuso** [comment]
@Vyom-Yadav can we mark this as ready for review?

**Comment by Vyom-Yadav** [comment]
> @Vyom-Yadav can we mark this as ready for review?

I marked it as ready for review but coverage for FullIdent is in #11231, we would have to merge that first or add suppressions here to make CI green.

**Comment by rnveach** [comment]
> coverage for FullIdent is in https://github.com/checkstyle/checkstyle/pull/11231 , we would have to merge that first or add suppressions here to make CI green.

**Comment by romani** [comment]
>but coverage for FullIdent is in https://github.com/checkstyle/checkstyle/pull/11231,

Please do not block this PR by new Check implementation. I have big doubts on new check design. It will take a while to accept it.

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Comment by Vyom-Yadav** [comment]
> Please do not block this PR by new Check implementation. I have big doubts on new check design. It will take a while to accept it. This PR is much easier, let's finish it sooner.

Writing tests for covering FullIdent with VariableDeclarationDistanceUsage would be tedious and time-consuming. Think it would be better to use https://github.com/checkstyle/checkstyle/pull/11231 for complete coverage.

**Comment by Vyom-Yadav** [comment]
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/a99d8b4_2023120550/reports/diff/index.html

**Comment by rnveach** [comment]
@Vyom-Yadav Is this PR still blocked?

**Comment by Vyom-Yadav** [comment]
> @Vyom-Yadav Is this PR still blocked?

Yes, coverage is in #11231, see https://github.com/checkstyle/checkstyle/pull/11672#issuecomment-1454706508

**Comment by romani** [comment]
I do not understand how new Check creation can fix problem in DistanceUsage module.
All modules should be completely independent.

@Vyom-Yadav , please explain

**Comment by Vyom-Yadav** [comment]
> I do not understand how new Check creation can fix problem in DistanceUsage module.
All modules should be completely independent.

Both `VariableDeclarationDistanceUsage` and `ChainedMethodCallWrap` use `FullIdent`.
`FullIdent` is modified to now return the name of all types of method calls (with typecasting, generics, etc.).

typecasting, generics, etc., these code examples for coverage purposes are present in `ChainedMethodCallWrap`, creating new test cases for coverage of `FullIdent` with `VariableDeclarationDistanceUsage` is both tedious and time-consuming.

Complete coverage is already present in `ChainedMethodCallWrap`, that is why:

1. Merge `ChainedMethodCallWrap`
2. Merge this PR.

**Comment by romani** [comment]
@nrmancuso  and @rnveach 

we have hard time to split dependency of this PR from new Check, New Check is blocked by me as I have concern on design.
I think we can allow @Vyom-Yadav to leak in our pitest supprssions some new survivals in this PR on FullIdent and resolve pitest survivals in New Check PR - https://github.com/checkstyle/checkstyle/pull/11231 .

it will be sort of gambit to unblock PRs and easy moving forward.

**Comment by nrmancuso** [comment]
@Vyom-Yadav please ping me for review when CI is happy and you are ready

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Comment by Vyom-Yadav** [comment]
@nrmancuso How can we suppress sonar violation, have suppressed the same given by cs.

**Comment by nrmancuso** [comment]
> @nrmancuso How can we suppress sonar violation, have suppressed the same given by cs.

From (link to sonar dashboard): https://sonarcloud.io/project/issues?open=AYelm2J05TGpmZFw3Omw&id=org.checkstyle%3Acheckstyle

>`Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed.`

I am inclined to agree with Sonar and Checkstyle on this one; I am willing to extend technical debt in pitest temporarily, but this method is getting very difficult to understand, IMO

File: pom.xml
Comment (romani) [inline pom.xml:4598]: I think it is ok, such numbers have less role after we introduced survival reports.

**Comment by romani** [comment]
> Cognitive Complexity 

is not easy .... I tried. Such a complicated logic.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:415]: this method does not actually do only log violation.
lets make code `if (isViolation(....)) { log(....);}`

**Review by romani — CHANGES_REQUESTED** [review]
items:

**Comment by romani** [comment]
@Vyom-Yadav , I was able to fix CS violation, in extra commit, please consider it.

**Comment by Vyom-Yadav** [comment]
> Sonar was not cheated.
I think we can remove result from while expression and make break as we set false . If any complains on method size we can shorten some comments or expression.

Sonar still complains. See comment below.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:415]: @Vyom-Yadav , please do

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java]: Done.

**Comment by Vyom-Yadav** [comment]
Github, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java
Comment (Vyom-Yadav) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:492]: @romani Sonar can be cheated by extracting this condition.
```suggestion
        while (cannotThinkOfAResonableName(result, isUsedVariableDeclarationFound, currentSiblingAst)) {
```

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:492]: Please apply cheating update

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:492]: Difficult to name this condition, how about:
```java

    while (!isInitializationSequenceFound(result, isUsedVariableDeclarationFound, currentSiblingAst)) {

....

    private static boolean isInitializationSequenceFound(boolean result, boolean isUsedVariableDeclarationFound, DetailAST currentSiblingAst) {
        return !result
                || isUsedVariableDeclarationFound
                || currentSiblingAst == null;
    }
```

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:492]: @Vyom-Yadav won't some check fail on boolean method parameters?

**Comment by romani** [comment]
I am leaning to unusual decision: 
to create separate issue to optimize this method (fix later with smaller diff), and suppress Sonar. 

This PR blocks too much and too big. We need to split it.

@nrmancuso , @rnveach , please let us know if you ok with this.

**Comment by nrmancuso** [comment]
I am fine, let’s keep moving forward

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:492]: > @Vyom-Yadav won't some check fail on boolean method parameters?

`mvn clean verify` passed on local, don't know about sonar and other lints.

**Comment by romani** [comment]
@Vyom-Yadav , please squash third and second commits.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:492]: lets not overhack code, keep it as is

**Comment by Vyom-Yadav** [comment]
> @Vyom-Yadav , please squash third and second commits.

Done.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceTryResources.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceTryResources.java:108]: this is not expected, there is no reasonable way to move it closer.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceRegExp.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceRegExp.java:471]: why it becomes ok ?
because usage is block of similar statements ?

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceTryResources.java
Comment (Vyom-Yadav) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceTryResources.java:108]: Hmm, I see. The problem here is this an already existing bug didn't pop up due to another bug shadowing it.

Just for this expression, there are two bugs:

1. The calculated distance is 2.
2. This is an initialization sequence.

The no. `2` bug is fixed in this PR. **This is not an initialization sequence**. Due to that bug being fixed, now this is popping up as a violation. What do you think about adding an `until..` comment here and opening a separate issue for bug no `1`?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceRegExp.java
Comment (Vyom-Yadav) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceRegExp.java:471]: Yes, it is natural that the user doesn't want to break the streak of `isNull` just to get rid of the violation. The check is smart enough to understand this.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceTryResources.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceTryResources.java:108]: I am ok split this to separate issue.

Comment (Vyom-Yadav) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceTryResources.java:110]: Created https://github.com/checkstyle/checkstyle/issues/13154

**Comment by Vyom-Yadav** [comment]
Github, generate report

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceTryResources.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceTryResources.java:108]: I do not think we can do this to our users, it will block them to use new version, so kind of kill whole purpose of this PR (to improve this Check).
I can agree on extension of suppression and some false negatives ...... but not on false positives.

**Comment by romani** [comment]
@Vyom-Yadav , please rebase.

**Pull request closed without merging**
